Fix `-Zregparm` for LLVM builtins

winstonallo opened the pull request:
<!-- homu-ignore:start -->
<!--
    r? @tgross35 
-->
<!-- homu-ignore:end -->
This fixes the issue where `-Zregparm=N` was not working correctly when calling LLVM intrinsics

By default on `x86-32`, arguments are passed on the stack. The `-Zregparm=N` flag allows the first `N` arguments to be passed in registers instead. 

When calling intrinsics like `memset`, LLVM still passes parameters on the stack, which prevents optimizations like tail calls.

As proposed by @tgross35, I fixed this by setting the `NumRegisterParameters` LLVM module flag to `N` when the `-Zregparm=N` is set. 

```rust
// compiler/rust_codegen_llvm/src/context.rs#375-382
if let Some(regparm_count) = sess.opts.unstable_opts.regparm {
    llvm::add_module_flag_u32(
        llmod,
        llvm::ModuleFlagMergeBehavior::Error,
        "NumRegisterParameters",
        regparm_count,
    );
}
```
[Here](https://rust.godbolt.org/z/YMezreo48) is a before/after compiler explorer.

Here is the final result for the code snippet in the original issue:
```asm
entrypoint:
        push    esi
        mov     esi, eax
        mov     eax, ecx
        mov     ecx, esi
        pop     esi
        jmp     memset   ; Tail call parameters in registers
```

Fixes: https://github.com/rust-lang/rust/issues/145271

winstonallo commented:
I think this should fix it already, working on the tests now!

winstonallo commented:
I added a test verifying that the `memset` parameters are passed in registers as discussed.

nikic commented on tests/assembly-llvm/regparm-module-flag.rs:
```diff
@@ -0,0 +1,21 @@
+//@ assembly-output: emit-asm
+//@ compile-flags: -O -Zregparm=3 -Cunsafe-allow-abi-mismatch=regparm
+//@ only-x86
```
Can you please make this a minicore test instead?

winstonallo replied:
Just did that, still having the PIC issue unless I add `-C relocation-model=static` to the compile-flags. This is my `c_void` definition in minicore, did I miss something?
```rust
#[lang = "c_void"]
#[repr(u8)]
pub enum c_void {
    __variant1,
    __variant2,
}
```

petrochenkov commented:
r? nikic

winstonallo commented on tests/assembly-llvm/regparm-module-flag.rs:
```diff
@@ -0,0 +1,28 @@
+
+type c_void = u8;
+
+unsafe extern "C" {
+    fn memset(p: *mut c_void, val: i32, len: usize) -> *mut c_void;
+}
```
Will do in a bit:)

winstonallo replied:
Done:)

tgross35 reviewed (commented):
Thank you for doing a quick fix here!

Just a few requests to make sure we're testing what we expect

winstonallo commented on tests/assembly-llvm/regparm-module-flag.rs:
```diff
@@ -0,0 +1,28 @@
+//@ assembly-output: emit-asm
+//@ compile-flags: -O -Zregparm=3 --target=i686-unknown-linux-gnu -C relocation-model=static
+//@ needs-llvm-components: x86
+
```
I added those, however the tail call optimization does not happen for `N ∈ {1, 2}`, since at least one argument has to be pushed onto the stack and therefore cleaned up after the function returns (at least that is my understanding).

Would this kind of assertion be okay?
```rust
// REGPARM2-LABEL: entrypoint
// REGPARM2: movl %edx, %ecx
// REGPARM2: movl 16(%esp), %edx
// REGPARM2: calll memset
```

winstonallo replied:
This seems to be expected with LLVM, I get the same behavior for C code compiled with clang (TCO for `N = 3`, no TCO for `N ∈ {1, 2}`)

winstonallo replied:
✅

winstonallo commented on tests/assembly-llvm/regparm-module-flag.rs:
```diff
@@ -0,0 +1,28 @@
+}
+
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn entrypoint(len: usize, val: i32, ptr: *mut c_void) -> *mut c_void {
```
👍🏻

winstonallo commented on tests/assembly-llvm/regparm-module-flag.rs:
```diff
@@ -0,0 +1,28 @@
+
+// CHECK: mov{{.*}}eax
+// CHECK: mov{{.*}}ecx
+// CHECK: mov{{.*}}esi
```
✅

tgross35 reviewed (approved):
Two small nits but this looks good to me once CI passes, please squash

@nikic anything else?

winstonallo commented on tests/assembly-llvm/regparm-module-flag.rs:
```diff
@@ -0,0 +1,62 @@
+    // REGPARM3: movl %e{{.*}}, %esi
+    // REGPARM3: movl %e{{.*}}, %eax
+    // REGPARM3: movl %e{{.*}}, %ecx
+    // REGPARM3: jmp memset
```
`REGPARM2: mov %e{{..}}, %ecx` does not happen for `N = 2`, the asm looks like this:
```asm
	movl	%edx, %esi
	movl	16(%esp), %edx
	movl	%eax, (%esp)
	movl	%esi, %eax
	calll	memset
	movl	%esi, %eax
	addl	$8, %esp
```
I added an assertion for `movl	16(%esp), %edx`

winstonallo commented:
I squashed, let me know if there is anything else!

winstonallo committed:
Set NumRegisterParameters LLVM module flag to `N` when `-Zregparm=N` is
set

* Enforce the `-Zregparm=N` flag by setting the NumRegisterParameters
LLVM module flag * Add assembly tests verifying that the parameters are
passed in registers for reparm values 1, 2, and 3, for both LLVM
intrinsics and non-builtin functions * Add c_void type to minicore

bors commented:
:pushpin: Commit 02d4666fd2eb57803a6c3e7b6d4f27172273163e has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 02d4666fd2eb57803a6c3e7b6d4f27172273163e -->
<!-- homu: {"type":"Approved","sha":"02d4666fd2eb57803a6c3e7b6d4f27172273163e","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

winstonallo commented:
> Looks great! Thanks for making this happen, and welcome :)
Thank you very much for all the help! I'll keep an eye on the issue tracker, but if you come across anything you think would be a good fit, please let me know!

winstonallo commented:
Should I create a separate issue for the asm tests, or just submit a PR when I'm done?

klensy commented on tests/assembly-llvm/regparm-module-flag.rs:
```diff
@@ -0,0 +1,70 @@
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn entrypoint(len: usize, ptr: *mut c_void, val: i32) -> *mut c_void {
+    // REGPARM1-LABEL: entrypoint
+    // REGPARM1: movl %e{{.*}}, %ecx
+    // REPGARM1: pushl
+    // REPGARM1: pushl
+    // REGPARM1: calll memset
```
> Oops, can I still change that now that the PR is in a rollup?

No, but you can try to ask to stop rollup, or fixup in next pr.

winstonallo replied:
Okay, I'm still at work right now, will submit a fix later. Thank you for finding that!

winstonallo commented:
Fix the typo, should I call bors myself or does a maintainer have to do it?

RalfJung commented:
If we do this, do we have to *also* keep the logic in our own adjustments or does setting just the LLVM flag suffice?
